docs/extras/casts.md

```diff
@@ -0,0 +1,201 @@
+
+```
+⊢ @eq.{1} real
+    (@coe.{1 1} int real
+       (@coe_to_lift.{1 1} int real
+          (@coe_base.{1 1} int real
+	  ...
+```
+
+and the very first line tells you that the goal is an equality between two real numbers (and the second line indicates that the left hand side of this equality is a coercion from the integers).
+
+# "Special" coercions.
+
+The coercion from `ℕ` to `ℤ` is not defined as the one coming from the fact that `ℤ` is a ring; it is more computationally efficient to use the constructor. The two functions are equal, but this is a theorem and not a definition. The coercion from `ℚ` to `ℝ` is noncomputable (although it doesn't have to be), and the coercion from `ℝ` to `ℂ` is defined by hand rather than being part of a general scheme of maps from `ℝ` to lots of places. 
+
+# Avoiding `simp`.
+
+If you do want to prove "trivial in maths2 goals "by hand" for some reason, then you can write `set_option trace.simplify.rewrite true` before your theorem, see what `simp` is doing, and then mimic it.
+
+```lean
+import data.complex.basic
+
+example (a b : ℤ) : ((a + b : ℤ) : ℝ) = a + b := int.cast_add a b
+
+-- set_option trace.simplify.rewrite true
+example (a : ℤ) (b : ℕ) (c : ℚ) (d : ℝ) :
+(a : ℂ) + b * c - d = (((a + ((b * c) : ℚ) : ℚ) - d) : ℝ) :=
+begin
+  rw rat.cast_add,
+  rw rat.cast_coe_int,
+  rw complex.of_real_sub,
+  rw complex.of_real_add,
+  rw rat.cast_mul,
+  rw complex.of_real_mul,
+  rw rat.cast_coe_nat,
+  rw complex.of_real_int_cast,
+  rw complex.of_real_nat_cast,
+  rw complex.of_real_rat_cast,
+end
+```
+
+The second example shows the difficulties that you face when trying to do this by hand. Some of the lemmas are of this form:
+
+```
+rat.cast_add : ∀ (m n : ℚ), ↑(m + n) = ↑m + ↑n
+```
+
+This lemma is in the `rat` namespace, and it says that if you are coercing *from* `rat` to somewhere else, then adding and then coercing is the same as coercing and then adding.
+
+But others are of this form:
+
+```lean
+complex.of_real_mul : ∀ (r s : ℝ), ↑(r * s) = ↑r * ↑s
+```
+
+This lemma is in the `complex` namespace, but it is about coercions *to* the complexes from the reals. Writing code by hand is full of pitfalls like this? Why is this? Is this just poor design? No, it's much more complicated than this. The problem is that the coercions from `nat` to `int`, and from `real` to `complex` are *not* defined as special cases of generic coercions from `nat` to a general semiring or from `real` to a general complete field -- mathlib's designers made a conscious decision to use other more computationally efficient coercions, hence the confusion. The designers are not bothered by this, because `simp` works!
+
+Hopefully the example I gave above was sufficiently convoluted that if you really want to know the exact names of the lemmas you need, you can start to guess. Here's a brief selection. Because there are so many special cases, this can be a bit of a minefield.
+
+```
```

Comment:
Do you intend to start a code listing here and only close it at the end of the file?